lib/widgets/login/browser_login.dart

```diff
@@ -0,0 +1,181 @@
+    final int userId;
+    final Uri realm;
+    try {
+      if (uri.queryParameters case {
+        'otp_encrypted_api_key': final String otpEncryptedApiKeyStr,
+        'email': final String emailStr,
+        'user_id': final String userIdStr,
+        'realm': final String realmStr,
+      }) {
+        if (otpEncryptedApiKeyStr.isEmpty || emailStr.isEmpty || userIdStr.isEmpty || realmStr.isEmpty) {
+          throw 'Got invalid query params from browser redirect url';
+        }
+        otpEncryptedApiKey = otpEncryptedApiKeyStr;
+        realm = Uri.parse(realmStr);
+        userId = int.parse(userIdStr);
+        email = emailStr;
+      } else {
+        throw 'Got invalid query params from browser redirect url';
+      }
+    } catch (e, st) {
+      // TODO: Log error to Sentry
+      debugLog('$e\n$st');
+      return;
+    }
+
+    // Get the previously temporarily stored otp & serverSettings.
+    final GetServerSettingsResult serverSettings;
+    final String apiKey;
+    try {
+      final otp = _tempAuthOtp[realm];
+      _tempAuthOtp.clear();
+      final settings = _tempServerSettings[realm];
+      _tempServerSettings.clear();
+      if (otp == null) {
+        throw 'Failed to find the previously generated mobile_auth_otp';
+      }
+      if (settings == null) {
+        // TODO: Maybe try refetching instead of error-ing out.
+        throw 'Failed to find the previously stored serverSettings';
+      }
+
+      // Decode the otp XOR-ed api key
+      apiKey = _decodeApiKey(otp, otpEncryptedApiKey);
+      serverSettings = settings;
+    } catch (e, st) {
+      // TODO: Log error to Sentry
+      debugLog('$e\n$st');
+      return;
+    }
+
+    // TODO(#108): give feedback to user on SQL exception, like dupe realm+user
+    final accountId = await globalStore.insertAccount(AccountsCompanion.insert(
+      realmUrl: serverSettings.realmUri,
+      email: email,
+      apiKey: apiKey,
+      userId: userId,
+      zulipFeatureLevel: serverSettings.zulipFeatureLevel,
+      zulipVersion: serverSettings.zulipVersion,
+      zulipMergeBase: Value(serverSettings.zulipMergeBase),
+    ));
+
+    if (!context.mounted) {
+      return;
+    }
+    navigatorKey.currentState?.pushAndRemoveUntil(
+      HomePage.buildRoute(accountId: accountId),
+      (route) => (route is! LoginSequenceRoute),
+    );
+  }
+}
+
+/// Generates a `mobile_flow_otp` to be used by the server for
+/// mobile login flow, server XOR's the api key with the otp hex
+/// and returns the resulting value. So, the same otp that was passed
+/// to the server can be used again to decode the actual api key.
+String _generateMobileFlowOtp() {
+  final rand = Random.secure();
+  return hex.encode(rand.nextBytes(32));
+}
+
+String _decodeApiKey(String otp, String otpEncryptedApiKey) {
+  final otpHex = hex.decode(otp);
+  final otpEncryptedApiKeyHex = hex.decode(otpEncryptedApiKey);
+  return String.fromCharCodes(otpHex ^ otpEncryptedApiKeyHex);
```

Comment:
The `^` extension makes it kind of opaque to see what this is doing. If we were writing code that was dense with this sort of operation, that could be worth it; but there's just one of these, so let's expand it out to something more explicit:
```suggestion
  if (otpHex.length != otpEncryptedApiKeyHex.length) throw …;
  return String.fromCharCodes(Iterable.generate(otpHex.length,
    (i) => otpHex[i] ^ otpEncryptedApiKeyHex[i]));
```